### PR TITLE
migrate snooper tests to use new local DNS server

### PR DESCRIPTION
**Pull request by akarpz**
### Motivation

https://datadoghq.atlassian.net/browse/NPM-3012
https://datadoghq.atlassian.net/browse/NPM-3071
https://datadoghq.atlassian.net/browse/NPM-3044

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/tracer/testutil/testdns/test_dns_server.go [hunk 1]
```diff
@@ -20,20 +21,33 @@ import (
...
-		globalServer.Start("tcp")
-		globalServer.Start("udp")
+		globalServer.Start("tcp", port)
+		globalServer.Start("udp", port)
 	})
 
 	require.NoError(t, globalServerError)
-	return net.ParseIP("127.0.0.1")
+	return net.ParseIP(localhostAddr)
 }
```
Comment (leeavital) [inline pkg/network/tracer/testutil/testdns/test_dns_server.go:51]: I think this could be cleaner if GetServerIP had it's original signature (taking a `*testing.T`) and started both the non-port-53 server and the port-53 server.

File: pkg/network/tracer/tracer_linux_test.go [hunk 2]
```diff
@@ -71,9 +71,9 @@ func doDNSQuery(t *testing.T, domain string, serverIP string) (*net.UDPAddr, *ne
 	dnsClient := new(dns.Client)
 	dnsConn, err := dnsClient.Dial(dnsServerAddr.String())
 	require.NoError(t, err)
-	defer dnsConn.Close()
 	dnsClientAddr := dnsConn.LocalAddr().(*net.UDPAddr)
 	_, _, err = dnsClient.ExchangeWithConn(queryMsg, dnsConn)
+	_ = dnsConn.Close()
```
Comment (leeavital) [inline pkg/network/tracer/tracer_linux_test.go:76]: why this change?

**Review by leeavital — APPROVED** [review]

File: pkg/network/tracer/tracer_linux_test.go (hunk 2 above)
Comment (akarpz) [reply pkg/network/tracer/tracer_linux_test.go:76]: This suppresses an IDE complaint that we're not dealing with the error.. the other option is to wrap this in a closure which I feel is overkill for this use case. `defer` is nice, but we also are closing this connection very close to when we open it anyway

File: pkg/network/tracer/testutil/testdns/test_dns_server.go (hunk 1 above)
Comment (akarpz) [reply pkg/network/tracer/testutil/testdns/test_dns_server.go:51]: yeah that's fine... I think the only thing there is that the non-53 server will be started always, when it's only used in one test

File: pkg/network/dns/snooper_test.go [hunk 3]
```diff
@@ -343,13 +336,14 @@ func TestDNSOverNonPort53(t *testing.T) {
 	statKeeper := reverseDNS.statKeeper
 
 	domains := []string{
-		"nonexistent.com.net",
+		"nonexistent.net.com",
 	}
-	h := &handler{}
-	shutdown, port := newTestServer(t, localhost, 0, "udp", h.ServeDNS)
-	defer shutdown()
+	ln, _ := net.Listen("tcp", "127.0.0.1:0")
```
Comment (hmahmood) [inline pkg/network/dns/snooper_test.go:341]: For the cases where we need a random port, I think we should just spin up a server and shut it down in the test. 

We can also pass in `0` for the port for the server spin up code to pick a port.

File: pkg/network/tracer/testutil/testdns/test_dns_server.go [hunk 4]
```diff
@@ -20,20 +21,34 @@ import (
...
 
 // GetServerIP returns the IP address of the test DNS server. The test DNS server returns canned responses for several
 // known domains that are used in integration tests.
 //
 // see server#start to see which domains are handled.
-func GetServerIP(t *testing.T) net.IP {
+func GetServerIP(t *testing.T, port int) net.IP {
+	if port != 53 {
+		non53ServerOnce.Do(func() {
+			globalServer.Start("tcp", port)
```
Comment (hmahmood) [inline pkg/network/tracer/testutil/testdns/test_dns_server.go:40]: How does this work if the dns server with port 53 is already running since you're using the same `globalServer` variable for both?

File: pkg/network/dns/snooper_test.go (hunk 3 above)
Comment (akarpz) [reply pkg/network/dns/snooper_test.go:341]: that is what we were doing before... I think the idea here was to centralize the server code. but I was on the fence when doing this, wdyt @leeavital

Comment (leeavital) [reply pkg/network/dns/snooper_test.go:341]: I don't have strong feelings, if we prefer to have a server-per-test, I'm fine with it. My thinking was it was nice to have a single global test server because:
* tests run faster when they don't each have to spin up a server
* tests don't have to deal with managing the lifecycle of 
* the nature of a stubbed server makes having a global easy
* (obsolete) it was unlikely for the global server to conflict with other test fixtures since it listened on a nonsense ip (10.10.10.10)

Comment (hmahmood) [reply pkg/network/dns/snooper_test.go:341]: I was only proposing a local server for this test where we don't use port 53. The other tests can use the global server. In addition, the server creation/run code seems to be using the same var, `globalServer`, for both servers, which is likely going to be brittle.

File: pkg/network/tracer/testutil/testdns/test_dns_server.go (hunk 4 above)
Comment (leeavital) [reply pkg/network/tracer/testutil/testdns/test_dns_server.go:40]: The constructor `newServer()` used to take care of creating interfaces, and no longer does. We can dispense with storing the `server`  in a global variable completely now that the server uses loopback

**Review by hmahmood — DISMISSED** [review]

File: pkg/network/tracer/testutil/testdns/test_dns_server.go [hunk 5]
```diff
@@ -10,30 +10,33 @@ package testdns
...
 var serverOnce sync.Once
 
+const localhostAddr = "127.0.0.1"
+
 // GetServerIP returns the IP address of the test DNS server. The test DNS server returns canned responses for several
 // known domains that are used in integration tests.
 //
 // see server#start to see which domains are handled.
 func GetServerIP(t *testing.T) net.IP {
+	var err error
```
Comment (leeavital) [inline pkg/network/tracer/testutil/testdns/test_dns_server.go:30]: What will happen if:

* test #1 calls GetServerIP and the sync.Once function fails 
* test #2 calls GetServerIP
?

Comment (akarpz) [reply pkg/network/tracer/testutil/testdns/test_dns_server.go:30]: I see what you mean. I think we need to keep the globalError and have it set by the call to `ListenAndServe()`

**Review by leeavital — DISMISSED** [review]

**Review by leeavital — APPROVED** [review]
